packages/server/graphql/queries/helpers/fetchAllRepoIntegrations.ts

```diff
@@ -0,0 +1,66 @@
+import {getPermsByTaskService, getPrevRepoIntegrations} from './repoIntegrationHelpers'
+import JiraProjectId from 'parabol-client/shared/gqlIds/JiraProjectId'
+import {GQLContext} from '../../graphql'
+import {GraphQLResolveInfo} from 'graphql'
+import fetchGitHubRepos from './fetchGitHubRepos'
+import {JiraProject} from 'parabol-client/utils/AtlassianManager'
+import IntegrationRepoId from 'parabol-client/shared/gqlIds/IntegrationRepoId'
+
+const fetchAllRepoIntegrations = async (
+  teamId: string,
+  userId: string,
+  context: GQLContext,
+  info: GraphQLResolveInfo
+) => {
+  const {dataLoader} = context
+  const permLookup = await getPermsByTaskService(dataLoader, teamId, userId)
+  // get prevRepoIntegrations to determine when the repo was last used
+  const [prevRepoIntegrations, jiraProjectsRes, githubRepos] = await Promise.all([
+    getPrevRepoIntegrations(userId, teamId, permLookup),
+    dataLoader.get('fetchAtlassianProjects').load({teamId, userId}),
+    fetchGitHubRepos(teamId, userId, dataLoader, context, info)
+  ])
+  const jiraProjects = jiraProjectsRes.map((project: JiraProject & {cloudId: string}) => ({
+    ...project,
+    id: JiraProjectId.join(project.cloudId, project.key),
+    teamId,
+    userId
+  }))
+
+  const allRepoIntegrations = [...jiraProjects, ...githubRepos]
+  const prevRepoIntegrationsIds = new Set<string>()
+  prevRepoIntegrations.forEach((integration) => {
+    const integrationId = IntegrationRepoId.join(integration)
+    if (!integrationId) return
+    prevRepoIntegrationsIds.add(integrationId)
+  })
+  // always have lastUsedAt be a Date (to make the sort easier below)
+  const repos = allRepoIntegrations.map((repo) => {
+    if (prevRepoIntegrationsIds.has(repo.id)) {
+      const prevRepoIntegration = prevRepoIntegrations.find((repoIntegration) => {
+        const integrationId = IntegrationRepoId.join(repoIntegration)
+        return repo.id === integrationId
+      })
+      return {
```

Comment:
-1 isn't this function supposed to filter out integrations that exist in prevRepoIntegrations?

Reply:
No, this function is fetching the `prevRepoIntegrations` because the `lastUsedAt` field doesn't exist on the Jira projects or GH repos. So, I'm getting all of the ids of previous integrations for that user & team, and then, when mapping through the `repoIntegrations`, I check whether the `repoIntegration` exists in `prevRepoIntegrationsIds`. If it does, I use the `lastUsedAt` date, if not, I create a new one with `new Date(0)`.

I have a comment [here](https://github.com/ParabolInc/parabol/blob/be868d328092806995a27002d1ab696c1d776181/packages/server/graphql/queries/helpers/fetchAllRepoIntegrations.ts#L17) that tries to explain this, but I'd be happy to add more info if it's unclear.

Reply:
ohhhh! i think i was stuck in the old way of thinking. this makes sense!
we can remove that extra find by storing the lastUsedAt in a lookup table instead of using the Set

For your consideration:


```ts
  const fetchedRepoIntegrations = [...jiraProjects, ...githubRepos]
  const repoIntegrationsLastUsedAt = {} as {[repoIntegrationId: string]: Date}
  prevRepoIntegrations.forEach((integration) => {
    const integrationId = IntegrationRepoId.join(integration)
    if (!integrationId) return
    repoIntegrationsLastUsedAt[integrationId] = integration.lastUsedAt
  })
  return fetchedRepoIntegrations
    .map((repo) => ({
      ...repo,
      // always have lastUsedAt be a Date (to make the sort easier below)
      lastUsedAt: repoIntegrationsLastUsedAt[repo.id] ?? new Date(0)
    }))
    .sort((a, b) =>
      a.lastUsedAt > b.lastUsedAt
        ? -1
        : a.service < b.service
        ? -1
        : a.id.toLowerCase() < b.id.toLowerCase()
        ? -1
        : 1
    )
```